Script to Convert AWS Credentials to C-style Strings

AGlass0fMilk opened the pull request:
Created script to automate conversion of AWS credential files to C-style strings

Usage is documented by the Python argparser. Briefly, you can save all the credential files from AWS to a single directory. You then execute this script (eg: `python aws-converter.py ./test-certs`) and give the directory containing the credential files. The script will then generate a header file (given by --output-file flag, or defaults to `aws_credentials.h`) based on the current template.

In the future this can be extended or serve as a starting point for others to develop their own tooling for deploying AWS certs. It eliminates a bunch of (potentially repetitive) copy-pasting and associated headaches.

AGlass0fMilk commented:
Directions should be updated to mention the usage of this script after #8 is merged.

LDong-Arm commented on aws-cert-converter.py:
```diff
@@ -0,0 +1,132 @@
+    ' * PEM-encoded root CA certificate\n'
+    ' *\n'
+    ' * Must include the PEM header and footer,\n'
+    ' * and every line of the body needs to be quoted and end with \\n:\n'
+    ' * "-----BEGIN CERTIFICATE-----\\n"\n'
+    ' * "...base64 data...\\n"\n'
+    ' * "-----END CERTIFICATE-----";\n'
+    ' */\n'
+    'const char rootCA[] = \n'
+    '{0};\n'
+    '\n'
+    '/*\n'
+    ' * PEM-encoded client certificate\n'
+    ' *\n'
+    ' * Must include the PEM header and footer,\n'
+    ' * and every line of the body needs to be quoted and end with \\n:\n'
+    ' * "-----BEGIN CERTIFICATE-----\\n"\n'
+    ' * "...base64 data...\\n"\n'
+    ' * "-----END CERTIFICATE-----";\n'
+    ' */\n'
+    'const char clientCrt[] = \n'
+    '{1};\n'
+    '\n'
+    '/*\n'
+    ' * PEM-encoded client private key.\n'
+    ' *\n'
+    ' * Must include the PEM header and footer,\n'
+    ' * and every line of the body needs to be quoted and end with \\n:\n'
+    ' * "-----BEGIN RSA PRIVATE KEY-----\\n"\n'
+    ' * "...base64 data...\\n"\n'
+    ' * "-----END RSA PRIVATE KEY-----";\n'
+    ' */\n'
+    'const char clientKey[] = \n'
+    '{2};\n'
+    '\n'
+    '}}\n'
+    '}}\n'
+    '\n'
+    '#endif\n'
+)
```
Python's multiline strings make life a lot easier:
```suggestion
header_template = (
    """
// AWS Certificates

#ifndef AWS_CREDENTIALS_H
#define AWS_CREDENTIALS_H

namespace aws {{
namespace credentials {{
/*
    * PEM-encoded root CA certificate
    *
    * Must include the PEM header and footer,
    * and every line of the body needs to be quoted and end with \\n:
    * "-----BEGIN CERTIFICATE-----\\n"
    * "...base64 data...\\n"
    * "-----END CERTIFICATE-----";
    */
const char rootCA[] = 
{0};

/*
    * PEM-encoded client certificate
    *
    * Must include the PEM header and footer,
    * and every line of the body needs to be quoted and end with \\n:
    * "-----BEGIN CERTIFICATE-----\\n"
    * "...base64 data...\\n"
    * "-----END CERTIFICATE-----";
    */
const char clientCrt[] = 
{1};

/*
    * PEM-encoded client private key.
    *
    * Must include the PEM header and footer,
    * and every line of the body needs to be quoted and end with \\n:
    * "-----BEGIN RSA PRIVATE KEY-----\\n"
    * "...base64 data...\\n"
    * "-----END RSA PRIVATE KEY-----";
    */
const char clientKey[] = 
{2};

}}
}}

#endif
    """
)
```

LDong-Arm replied:
Also, we may not need comments/docs in the template since the generated file already contains the keys? And the format (e.g. `\n`) will be obvious to users if they want to replace any keys.

LDong-Arm commented on aws-cert-converter.py:
```diff
@@ -0,0 +1,132 @@
+import argparse
```
```suggestion
#!/usr/bin/env python

import argparse
```
The file can be committed with permission +x.

AGlass0fMilk commented:
@ithinuel @LDong-Arm any update on this?

LDong-Arm commented:
@AGlass0fMilk Thanks for following up.

What about taking a certificate (or just any text file) as an input, and outputting a C string (with `"` and `\n`added) to the screen? Then the user will be able to paste it into the code.

This kind of utility would be general purpose and not specific to this example. What do you think?

AGlass0fMilk commented:
@LDong-Arm Please review, I have integrated your changes and fixed a few other things. I had an extra `global` variable declaration in there that seemed to cause syntax error issues after I switched to using python3. That should be fixed now.

AGlass0fMilk commented:
@adbridge @LDong-Arm Any update on this?

AGlass0fMilk commented:
@LDong-Arm @adbridge Any update on this?

LDong-Arm commented:
@evedon Are we happy to get this in?

evedon commented:
> Directions should be updated to mention the usage of this script after #8 is merged.

I am fine with the script but let's get README updated with usage instructions in this PR.

AGlass0fMilk commented:
@evedon I have updated the README as requested. See 894eee2

LDong-Arm commented on README.md:
```diff
@@ -39,6 +39,16 @@ You can build the project with all supported [Mbed OS build tools](https://os.mb
+    
+    Then, you can run the script to automatically generate the necessary code from the credentials:
+    
+    `python aws-cert-converter.py aws-credentials`
```
```suggestion
    ```
    python aws-cert-converter.py aws-credentials
    ```
```

LDong-Arm reviewed (commented):
@AGlass0fMilk The doc looks good to me, I just suggested a small format change.

AGlass0fMilk commented:
@LDong-Arm I have applied your suggested change.

LDong-Arm reviewed (approved):
Thanks @AGlass0fMilk

evedon commented:
Thanks for the PR
